src/daemon/daemon.cpp

```diff
@@ -678,26 +763,48 @@ mp::Daemon::Daemon(std::unique_ptr<const DaemonConfig> the_config)
 {
     connect_rpc(daemon_rpc, *this);
     std::vector<std::string> invalid_specs;
-    bool mac_addr_missing{false};
+
     for (auto& entry : vm_instance_specs)
     {
         const auto& name = entry.first;
         auto& spec = entry.second;
 
+        // Store here the MAC addresses we need add to the daemon's set. But we'll add them only if this instance
+        // is not invalid.
+        std::unordered_set<std::string> instance_mac_addresses;
+
         if (!config->vault->has_record_for(name))
         {
             invalid_specs.push_back(name);
             continue;
         }
 
-        auto mac_addr = spec.mac_addr;
-        if (mac_addr.empty())
+        // Check MAC addresses of the default and extra interfaces. For each MAC, repetition on the instance MAC
+        // addresses and in the daemon's MAC addresses is checked. String validity was already checked in
+        // load_db().
+        if (!mpu::valid_mac_address(spec.default_interface.mac_address) ||
+            !instance_mac_addresses.insert(spec.default_interface.mac_address).second ||
+            allocated_mac_addrs.find(spec.default_interface.mac_address) != allocated_mac_addrs.end())
         {
-            mac_addr = mp::utils::generate_mac_address();
-            vm_instance_specs[name].mac_addr = mac_addr;
-            mac_addr_missing = true;
+            mpl::log(mpl::Level::warning, category,
+                     fmt::format("{} has invalid or repeated default MAC address {}", name,
+                                 spec.default_interface.mac_address));
+            invalid_specs.push_back(name);
+            continue;
+        }
+
+        for (auto& iface : spec.extra_interfaces)
+        {
+            if (!mp::utils::valid_mac_address(iface.mac_address) ||
+                !instance_mac_addresses.insert(iface.mac_address).second ||
+                allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end())
+            {
```

Comment:
Same points on leftover string checking and insert before repetition check (this one more relevant here). This could probably be replaced with (not compiled, tested, or formatted):

```suggestion
            if (allocated_mac_addrs.find(iface.mac_address) != allocated_mac_addrs.end() || 
                !instance_mac_addresses.insert(iface.mac_address).second)
            {
```

Reply:
Right, removed the validity checking.